### DISCUSSION TITLE
Aptos Keyless SDK How can Zod schemas be used to define and validate structured payloads

### DISCUSSION THREAD

**Discussion by BigOTF** [aptos-labs/aptos-developer-discussions#557] [Questions] [answered]
### Describe your question in detail.

 How can Zod schemas be used to define and validate structured payloads for ID tokens, including extended properties like user profiles and email information.

export const emailScopedPayloadSchema = nonceEncryptedIdTokenSchema.extend({
  email: z.string().optional(),
  email_verified: z.boolean(),
});

### Which operating system are you using?

Windows

### Which SDK or tool are you using? (if any)

TypeScript SDK

**Comment by Pcharlesme** [comment, marked as answer]
tried this approach  The nonceEncryptedIdTokenSchema extends idTokenSchema by adding a nonce field (string). 

```
export type IDToken = z.infer<typeof idTokenSchema>;

export type NonceEncryptedIdToken = z.infer<typeof nonceEncryptedIdTokenSchema>;

export type ProfileScopedPayloadSchema = z.infer<
  typeof profileScopedPayloadSchema
>;

export type EmailScopedPayloadSchema = z.infer<typeof emailScopedPayloadSchema>;

export type EncryptedScopedIdToken = z.infer<typeof scopedPayloadSchema>;
```

```
Import { z } from "zod";

export const idTokenSchema = z.object({
  aud: z.string(),
  exp: z.number(),
  iat: z.number(),
  iss: z.string(),
  sub: z.string(),
});

export const nonceEncryptedIdTokenSchema = idTokenSchema.extend({
  nonce: z.string(),
});

export const profileScopedPayloadSchema = nonceEncryptedIdTokenSchema.extend({
  family_name: z.string().optional(),
  given_name: z.string().optional(),
  locale: z.string().optional(),
  name: z.string(),
  picture: z.string().optional(),
});

export const emailScopedPayloadSchema = nonceEncryptedIdTokenSchema.extend({
  email: z.string().optional(),
  email_verified: z.boolean(),
});
```

**Reply by BigOTF** [reply]
so, what do i need to pass as function in the scopedPayloadSchema.parse

**Comment by Pcharlesme** [comment]
export type EncryptedScopedIdToken = z.infer<typeof scopedPayloadSchema>;

`scopedPayloadSchema.parse(scopedData);`

but your object should the method call

```
const scopedData = {
  ...profileData,
  email: "user@example.com",
  email_verified: true,
};
```

**Reply by BigOTF** [reply]
implemented it now and it works fine. Thank you
